contracts/v2/marketplace/TokenMarketplaceV2.sol

```diff
@@ -0,0 +1,509 @@
+    address indexed _currentOwner,
+    address indexed _bidder,
+    uint256 _amount
+  );
+
+  event AuctionEnabled(
+    uint256 indexed _tokenId,
+    address indexed _auctioneer
+  );
+
+  event AuctionDisabled(
+    uint256 indexed _tokenId,
+    address indexed _auctioneer
+  );
+
+  struct Offer {
+    address bidder;
+    uint256 offer;
+  }
+
+  struct Listing {
+    uint256 price;
+    address seller;
+  }
+
+  bool private _notEntered = true;
+
+  // Min increase in bid/list amount
+  uint256 public minBidAmount = 0.04 ether;
+
+  // Interface into the KODA world
+  IKODAV2Methods public kodaAddress;
+
+  // KO account which can receive commission
+  address public koCommissionAccount;
+
+  uint256 public artistRoyaltyPercentage = 100;
+  uint256 public platformFeePercentage = 25;
+
+  // Token ID to Offer mapping
+  mapping(uint256 => Offer) offers;
+
+  // Token ID to Listing
+  mapping(uint256 => Listing) public listings;
+
+  // Explicitly disable sales for specific tokens
+  mapping(uint256 => bool) disabledTokens;
+
+  ///////////////
+  // Modifiers //
+  ///////////////
+
+  modifier onlyWhenOfferOwner(uint256 _tokenId) {
+    require(offers[_tokenId].bidder == msg.sender, "Not offer maker");
+    _;
+  }
+
+  modifier onlyWhenTokenExists(uint256 _tokenId) {
+    require(kodaAddress.exists(_tokenId), "Token does not exist");
+    _;
+  }
+
+  modifier onlyWhenBidOverMinAmount(uint256 _tokenId) {
+    require(msg.value >= offers[_tokenId].offer.add(minBidAmount), "Offer not enough");
+    _;
+  }
+
+  modifier onlyWhenTokenAuctionEnabled(uint256 _tokenId) {
+    require(!disabledTokens[_tokenId], "Token not enabled for offers");
+    _;
+  }
+
+  modifier nonReentrant() {
+    // On the first call to nonReentrant, _notEntered will be true
+    require(_notEntered, "ReentrancyGuard: reentrant call");
+
+    // Any calls to nonReentrant after this point will fail
+    _notEntered = false;
+
+    _;
+
+    // By storing the original value once again, a refund is triggered (see
+    // https://eips.ethereum.org/EIPS/eip-2200)
+    _notEntered = true;
+  }
+
+  /////////////////
+  // Constructor //
+  /////////////////
+
+  // Set the caller as the default KO account
+  constructor(IKODAV2Methods _kodaAddress, address _koCommissionAccount) public {
+    kodaAddress = _kodaAddress;
+    koCommissionAccount = _koCommissionAccount;
+    super.addAddressToWhitelist(msg.sender);
+  }
+
+  //////////////////////////
+  // User Bidding Actions //
+  //////////////////////////
+
+  function placeBid(uint256 _tokenId)
+  public
+  payable
+  whenNotPaused
+  nonReentrant
+  onlyWhenTokenExists(_tokenId)
+  onlyWhenBidOverMinAmount(_tokenId)
+  onlyWhenTokenAuctionEnabled(_tokenId)
+  {
+    require(!isContract(msg.sender), "Unable to place a bid as a contract");
+    _refundHighestBidder(_tokenId);
+
+    offers[_tokenId] = Offer(msg.sender, msg.value);
+
+    address currentOwner = kodaAddress.ownerOf(_tokenId);
```

Comment:
if this is for the subgraph, can the subgraph not make the owner of call to save GAS? Just thinking too that this is only valid at the time of placing the bid and if someone transferred, then this piece of info will be immediately out of date. I see the new owner could accept the offer if they wanted which is fine in my mind

Reply:
I think looking up ownership at point of confirmation is the safest, however it is more costly as you say.